### DISCUSSION TITLE
Don't clear depth buffer after rendering

### DISCUSSION THREAD

**Discussion by FinleyPan** [google/filament#4451] [Q&A]
Hey. There are two views and one render target, I want to draw the 2nd views on top of the 1st view:
```cpp
filament::Scene *first_scene = nullptr;
filament::Scene *second_scene = nullptr;
filament::View* first_view     = nullptr;
filament::View* second_view    = nullptr;
filament::Renderer *renderer = nullptr;
filament::RenderTarget *target = nullptr;
//codes to initialize views, render and target
...

first_view->setScene(first_scene);
second_view->setScene(second_scene);
first_view->setRenderTarget(target);
second_view->setRenderTarget(target);
....
renderer->render(first_view);
renderer->render(second_view);
```
`target` is created with color and depth texture. I want to keep the depth buffer after the 1st rendering. However, I find filament always clear depth buffer in function [ FRenderer::initializeClearFlags](https://github.com/google/filament/blob/main/filament/src/Renderer.cpp#L1156-L1164):
```cpp
void FRenderer::initializeClearFlags() {
    // We always discard and clear the depth+stencil buffers -- we don't allow sharing these
    // across views (clear implies discard)
    mDiscardStartFlags = ((mClearOptions.discard || mClearOptions.clear) ?
                          TargetBufferFlags::COLOR : TargetBufferFlags::NONE)
                         | TargetBufferFlags::DEPTH_AND_STENCIL;

    mClearFlags = (mClearOptions.clear ? TargetBufferFlags::COLOR : TargetBufferFlags::NONE)
            | TargetBufferFlags::DEPTH_AND_STENCIL;
}
```
Is there a way to overcome it?

**Comment by romainguy** [comment]
@pixelflinger

**Reply by FinleyPan** [reply]
Hey @romainguy @pixelflinger I modify method `FRenderer::initializeClearFlags` as:
```cpp
void FRenderer::initializeClearFlags() {
    // We always discard and clear the depth+stencil buffers -- we don't allow sharing these
    // across views (clear implies discard)
    mDiscardStartFlags = ((mClearOptions.discard || mClearOptions.clear) ?
                          TargetBufferFlags::COLOR : TargetBufferFlags::NONE)
                         | TargetBufferFlags::STENCIL;

    mClearFlags = (mClearOptions.clear ? TargetBufferFlags::COLOR : TargetBufferFlags::NONE)
            | TargetBufferFlags::STENCIL;
}
```
Then it works out for my problem, but I don't know whether it will cause any bugs.

**Reply by pixelflinger** [reply]
This would impact performance, because the depth buffer would always be written to memory, when in fact it's almost never needed.

**Reply by FinleyPan** [reply]
> This would impact performance, because the depth buffer would always be written to memory, when in fact it's almost never needed.

Hi. Is the memory you say device(GPU) memory or host memory?

**Reply by romainguy** [reply]
We're talking about off-tile memory, which on mobile devices is unified memory shared between the GPU and CPU.

**Comment by pixelflinger** [comment]
I believe if you set the `SAMPLEABLE` flag on the depth texture of your custom rendertarget, we will preserve it.
This is not supported for non-custom render targets.

**Reply by FinleyPan** [reply]
> I believe if you set the `SAMPLEABLE` flag on the depth texture of your custom rendertarget, we will preserve it.
> This is not supported for non-custom render targets.

Do you mean sample the depth texture in material and then do something similar to early-z test?

**Reply by FinleyPan** [reply]
Does filament material support write to `gl_FragDepth`?

**Reply by romainguy** [reply]
You can do it yes. Not sure how it's related though? @pixelflinger was simply saying you should add the SAMPLEABLE flag to your target.

**Reply by pixelflinger** [reply]
> > I believe if you set the `SAMPLEABLE` flag on the depth texture of your custom rendertarget, we will preserve it.
> > This is not supported for non-custom render targets.
> 
> Do you mean sample the depth texture in material and then do something similar to early-z test?

No, just add the `SAMPLEABLE` flag to the depth texture's usage bit. When this flag is set, we won't discard the buffer.

**Reply by FinleyPan** [reply]
Hi, @pixelflinger I create my depth texture as you said:
```cpp
            depth_ = filament::Texture::Builder()
                .width(width)
                .height(height)
                .sampler(filament::Texture::Sampler::SAMPLER_2D)
                .format(filament::Texture::InternalFormat::DEPTH32F)
                .usage(filament::Texture::Usage::DEPTH_ATTACHMENT | filament::Texture::Usage::SAMPLEABLE)
                .build(*engine);
```
And I created a test in which there are two entities: a box and a triangle. If I put them into one scene and render once, the result is aspected:
![image](https://user-images.githubusercontent.com/26212839/128510650-d6367264-d2f9-4cf8-a87d-7a019c22a99a.png)
when I  put them into two scenes and render twice into the same target, the result is as follows:
![image](https://user-images.githubusercontent.com/26212839/128510884-bf3d843b-5e0a-421c-9ad2-f9a43718f612.png)
the problem seems not solved to me.

**Reply by pixelflinger** [reply]
Okay, I will look into it. I think this should have worked.
Can you share your sample app?

**Reply by FinleyPan** [reply]
Hi @pixelflinger I put the test [here](https://github.com/FinleyPan/filament/tree/dev). Please check it, thanks!   
By changing the [line](https://github.com/FinleyPan/filament/blob/dev/samples/multiple_fbos.cpp#L522):
```cpp
bool render_once = true;
```
you can control over rendering all entities once or twice in two different scenes(top is result of rendering twice, bottom is that of once):
<img width="380" alt="wecom-temp-208885b228b798fa145a14a9da129964" src="https://user-images.githubusercontent.com/26212839/128624218-dd404947-05b9-4ddc-a112-5b4a5a6750ce.png">
<img width="340" alt="wecom-temp-69e0f223b941e141f8da795fb4ef1db2" src="https://user-images.githubusercontent.com/26212839/128624223-2046d70c-a71e-4424-8f41-280a84428176.png">

**Reply by FinleyPan** [reply]
@pixelflinger Hey, is there any progress :)? I found a similar issue #3962, but the fix does not work for my case.

**Reply by pixelflinger** [reply]
Ah yes, I didn't realize that you where not redrawing the object each frame -- this use case is not supported right now.

What _should_ work is to have a several views using the same `RenderTarget` and have the depth buffer preserved (if marked as `SAMPLEABLE`) **within the same** `beginFrame`/`endFrame`. Currently buffers are cleared no matter what at the beginning of a frame.

I would file an "enhancement" issue at this point.

**Reply by FinleyPan** [reply]
> Ah yes, I didn't realize that you where not redrawing the object each frame -- this use case is not supported right now.  

Can you illustrate "redrawing object each frame" concretely？

**Reply by FinleyPan** [reply]
> What _should_ work is to have a several views using the same `RenderTarget` and have the depth buffer preserved (if marked as `SAMPLEABLE`) **within the same** `beginFrame`/`endFrame`. Currently buffers are cleared no matter what at the beginning of a frame.

I think my demo conforms to this. The only difference is times of calling `filament::Renderer::render` (once and twice) in each  `beginFrame`/`endFrame`

**Reply by FinleyPan** [reply]
Hey @pixelflinger, is there any updates?

**Comment by lucasleandro1805** [comment]
Hi, any fixes for that?
I`m using official AAR for android
